Protect ENI and Azure IPAM from misbehaving cloud APIs

tgraf opened the pull request:
See individual commits

qmonnet commented on pkg/ipam/node.go:
```diff
@@ -645,6 +650,15 @@ func (n *Node) updateLastResync(syncTime time.Time) {
+		}
+		return fmt.Errorf("instances API is unstable. Blocking mutating operations. See logs for details.")
+	}
+
```
This chunk makes unit tests fail. As far as I understand this is because API has never been sync'ed in the first place, and synchronisation is called at the end of this function (`n.manager.resyncTrigger.Trigger()`).

Adding `n.manager.resyncTrigger.Trigger()` just above the chunk works but is probably not the expected fix... It should more likely be part of `n.retry.Trigger()`? Maybe something like this, for the last chunk of this set in `pkg/ipam/node_manager.go`:

``` go
		retry, err := trigger.NewTrigger(trigger.Parameters{
			Name:        fmt.Sprintf("ipam-pool-maintainer-%s-retry", resource.Name),
			MinInterval: time.Minute, // large minimal interval to not retry too often
			TriggerFunc: func(reasons []string) {
				poolMaintainer.Trigger()
				n.resyncTrigger.Trigger() // NOT SURE IF CORRECT, BUT FIXES UNIT TESTS
			},
		})
		if err != nil {
			node.logger().WithError(err).Error("Unable to create ...")
			return false
		}
		node.retry = retry
```

tgraf committed:
eni/azure: Only sync IPAM state to CiliumNode resource if cloud API sync was successful

Cloud APIs can get into a bad state. This could result in the operator
being restarted. If that happens and the Cloud API synchronization then
failed the CiliumNode resource would have its status overwritten. This
is not desirable. Require a sucessful Cloud API sync before updating the
CiliumNode resource.

Fixes: #11052

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
eni/azure: Don't perform mutating operations while instances API is unstable

It is possible for the cloud APIs being used by the operator to get into
a state where POST and PATCH operations are still succeeding while GET
operations are failing. This can result in the operator to continously
creating resources while being unable to ever synchronize the state
successfully. Require a successful synchronization of all resources in
order to continue performing mutating operations.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
eni/azure: Fail if initial cloud API synchronization fails

The initial synchronization is blocking but did not return the error so
far. Treat the initial synchronization as critical. If that can't
succeed, restart the operator to indicate the problem.

IP allocation will not succeed anyway.

Signed-off-by: Thomas Graf <thomas@cilium.io>

christarazi commented on pkg/ipam/node.go:
```diff
@@ -653,6 +658,15 @@ func (n *Node) updateLastResync(syncTime time.Time) {
+		if n.retry != nil {
+			n.retry.Trigger()
+		}
+		return fmt.Errorf("instances API is unstable. Blocking mutating operations. See logs for details.")
```
Nit: we can use `errors.New()` for simple string errors. This can help reduce the usage and hopefully the need to import `fmt`, but I assume that we already use `fmt.Errorf` for good reasons in this file. Feel free to ignore if so.

tgraf replied:
I'm keeping `fmt.Errorf()` for now to keep it consistent with the rest of the code in this file. I think we can definitely start using `errors.New()` and `errors.Wrap()` but we should make a consistent decision to do so after some discussion because I'm sure it will have implications on assumptions that all contributors are currently making.

christarazi commented on pkg/ipam/node_manager.go:
```diff
@@ -181,9 +181,12 @@ func NewNodeManager(instancesAPI AllocationImplementation, k8sAPI CiliumNodeGett
-	n.instancesAPI.Resync(ctx)
+	resyncTime := n.instancesAPI.Resync(ctx)
+	if resyncTime.IsZero() {
+		return fmt.Errorf("Initial synchronization with instances API failed")
```
Nit: we can use `errors.New()` for simple string errors. This can help reduce the usage and hopefully the need to import `fmt`, but I assume that we already use `fmt.Errorf` for good reasons in this file. Feel free to ignore if so.

christarazi reviewed (approved):
Looks good to me! Have some non-blocking nits, feel free to skip.

ungureanuvladvictor commented on pkg/ipam/node_manager.go:
```diff
@@ -181,9 +181,12 @@ func NewNodeManager(instancesAPI AllocationImplementation, k8sAPI CiliumNodeGett
+func (n *NodeManager) Start(ctx context.Context) error {
 	// Trigger the initial resync in a blocking manner
-	n.instancesAPI.Resync(ctx)
+	resyncTime := n.instancesAPI.Resync(ctx)
```
I assume the `Resync` call stack at some point it will log the error from the cloud provider? If yes maybe add in the lower `fmt.Errorf` something like `"please check previous logs for errors"`?

ungureanuvladvictor reviewed (commented):
Just one q related to logging, otherwise will let current reviewers to handle this PR.
